File: Java/Floyd_Warshall.java
import java.util.Scanner;

public class FloydWarshall {
    final static int INF = 99999;

    // Function to print the solution matrix
    void printSolution(int dist[][], int V) {
        System.out.println("Shortest distances between every pair of vertices:");
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                if (dist[i][j] == INF)
                    System.out.print("INF" + "     ");
                else
                    System.out.print(dist[i][j] + "     ");
            }
            System.out.println();
        }
    }

    // Function to implement the Floyd-Warshall algorithm
    void floydWarshall(int graph[][], int V) {
        int dist[][] = new int[V][V];

        // Initialize the distance matrix with the input graph
        for (int i = 0; i < V; i++)
            for (int j = 0; j < V; j++)
                dist[i][j] = graph[i][j];

        // Update the distance matrix
        for (int k = 0; k < V; k++) {
            for (int i = 0; i < V; i++) {
                for (int j = 0; j < V; j++) {
                    if (dist[i][k] != INF && dist[k][j] != INF && dist[i][k] + dist[k][j] < dist[i][j]) {
                        dist[i][j] = dist[i][k] + dist[k][j];
                    }
                }
            }
        }

        printSolution(dist, V);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        FloydWarshall fw = new FloydWarshall();

        System.out.println("Enter the number of vertices:");
        int V = scanner.nextInt();  // User-defined number of vertices
        
        int[][] graph = new int[V][V];
        System.out.println("Enter the adjacency matrix (use " + INF + " for no direct path):");

        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                System.out.print("Distance from vertex " + i + " to vertex " + j + ": ");
                graph[i][j] = scanner.nextInt();
            }
        }

        fw.floydWarshall(graph, V);  // Pass V as a parameter
        scanner.close();
    }
}

File: go/FloydWarshall.go
// FloydWarshall.go
//
// Floyd–Warshall Algorithm - All-Pairs Shortest Path
//
// Description:
//   The Floyd–Warshall algorithm computes the shortest paths between
//   all pairs of vertices in a weighted directed graph. It can handle
//   negative edge weights (but not negative cycles).
//
// Purpose / Use cases:
//   - Network routing, transitive closure, and distance computation.
//   - Detect negative cycles in graphs.
//
// Approach / Methodology:
//   - Dynamic Programming approach.
//   - Initialize a distance matrix with given edge weights.
//   - Iteratively relax edges by trying all intermediate vertices.
//
// Complexity Analysis:
//   - Time: O(V³)
//   - Space: O(V²)
//
// File contents:
//   - FloydWarshall() function.
//   - PrintDistanceMatrix() utility.
//   - Example test in main().

package main

import (
	"fmt"
	"math"
)

// FloydWarshall computes all-pairs shortest paths.
// It modifies and returns the distance matrix.
func FloydWarshall(dist [][]float64) [][]float64 {
	n := len(dist)
	for k := 0; k < n; k++ {
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				if dist[i][k]+dist[k][j] < dist[i][j] {
					dist[i][j] = dist[i][k] + dist[k][j]
				}
			}
		}
	}
	return dist
}

// PrintDistanceMatrix prints a formatted distance matrix.
func PrintDistanceMatrix(dist [][]float64) {
	fmt.Println("All-Pairs Shortest Path Matrix:")
	for _, row := range dist {
		for _, val := range row {
			if val == math.Inf(1) {
				fmt.Printf("%4s ", "INF")
			} else {
				fmt.Printf("%4.0f ", val)
			}
		}
		fmt.Println()
	}
}

// main runs an example directed graph for Floyd–Warshall.
func main() {
	fmt.Println("Floyd–Warshall Algorithm Demo")

	const INF = math.Inf(1)
	dist := [][]float64{
		{0, 3, INF, 7},
		{8, 0, 2, INF},
		{5, INF, 0, 1},
		{2, INF, INF, 0},
	}

	result := FloydWarshall(dist)
	PrintDistanceMatrix(result)
}
